File: problem/Last_Day_Cross/latestDayToCross.js
/**
 * @param {number} row
 * @param {number} col
 * @param {number[][]} cells
 * @return {number}
 */
var latestDayToCross = function (row, col, cells) {
  const isPossible = (m, n, t) => {
    const grid = Array.from({ length: m + 1 }, () => new Uint8Array(n + 1));
    for (let i = 0; i < t; i++) {
      const [r, c] = cells[i];
      grid[r][c] = 1;
    }

    const qr = new Int32Array(m * n + 5);
    const qc = new Int32Array(m * n + 5);
    let head = 0, tail = 0;

    for (let c = 1; c <= n; c++) {
      if (grid[1][c] === 0) {
        qr[tail] = 1;
        qc[tail] = c;
        tail++;
        grid[1][c] = 1;
      }
    }

    const dr = [1, -1, 0, 0];
    const dc = [0, 0, 1, -1];

    while (head < tail) {
      const r = qr[head], c = qc[head];
      head++;
      for (let k = 0; k < 4; k++) {
        const nr = r + dr[k], nc = c + dc[k];
        if (nr > 0 && nc > 0 && nr <= m && nc <= n && grid[nr][nc] === 0) {
          grid[nr][nc] = 1;
          if (nr === m) return true;
          qr[tail] = nr;
          qc[tail] = nc;
          tail++;
        }
      }
    }
    return false;
  };

  let left = 0, right = row * col, ans = 0;
  while (left < right - 1) {
    const mid = left + ((right - left) >> 1);
    if (isPossible(row, col, mid)) {
      left = mid;
      ans = mid;
    } else {
      right = mid;
    }
  }
  return ans;
};

File: problem/Last_Day_Cross/latestDayToCross.c
#include <stdlib.h>
#include <string.h>

typedef struct
{
    int r, c;
} Node;

static int isPossible(int m, int n, int t, int **cells)
{
    int size = (m + 1) * (n + 1);
    unsigned char *grid = (unsigned char *)calloc(size, 1);
    if (!grid)
        return 0;

    for (int i = 0; i < t; i++)
    {
        int r = cells[i][0], c = cells[i][1];
        grid[r * (n + 1) + c] = 1;
    }

    Node *q = (Node *)malloc(sizeof(Node) * (m * n + 5));
    int head = 0, tail = 0;

    for (int c = 1; c <= n; c++)
    {
        if (grid[1 * (n + 1) + c] == 0)
        {
            q[tail++] = (Node){1, c};
            grid[1 * (n + 1) + c] = 1;
        }
    }

    int dr[4] = {1, -1, 0, 0};
    int dc[4] = {0, 0, 1, -1};

    while (head < tail)
    {
        Node cur = q[head++];
        for (int k = 0; k < 4; k++)
        {
            int nr = cur.r + dr[k], nc = cur.c + dc[k];
            if (nr > 0 && nc > 0 && nr <= m && nc <= n && grid[nr * (n + 1) + nc] == 0)
            {
                grid[nr * (n + 1) + nc] = 1;
                if (nr == m)
                {
                    free(q);
                    free(grid);
                    return 1;
                }
                q[tail++] = (Node){nr, nc};
            }
        }
    }

    free(q);
    free(grid);
    return 0;
}

int latestDayToCross(int row, int col, int **cells, int cellsSize, int *cellsColSize)
{
    int left = 0, right = row * col, ans = 0;
    while (left < right - 1)
    {
        int mid = left + (right - left) / 2;
        if (isPossible(row, col, mid, cells))
        {
            left = mid;
            ans = mid;
        }
        else
        {
            right = mid;
        }
    }
    return ans;
}